src/transformers/generation_utils.py

```diff
@@ -1879,27 +1916,20 @@ def contrastive_search(
                 if this_peer_finished_flag.item() == 0.0:
                     break
 
-            # prepare inputs
-            model_kwargs["use_cache"] = True
-            model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
-
             # if the first step in the loop, encode all the prefix and obtain three parameters: (1) past_key_values;
             # (2) last_hidden_states; (3) logit_for_next_step; (4) update model kwargs for the next step
             if model_kwargs.get("past") is None:
+
+                # prepare inputs
+                model_kwargs["use_cache"] = True
+                model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
+
                 # encode the given prefix and prepare model inputs; encoder-decoder model process the prefix and save
                 # the `encoder_outputs`
                 outputs = self(
                     **model_inputs, return_dict=True, output_hidden_states=True, output_attentions=output_attentions
                 )
 
-                # past_key_values is required for fast decoding
-                if "past_key_values" not in outputs:
```

Comment:
This check was moved below and made more general -- contrastive search is now compatible with other keys for `past`